lib/msf/core/db_manager/service.rb

```diff
@@ -131,15 +135,31 @@ def report_service(opts)
   end
 
   # Returns a list of all services in the database
-  def services(wspace = workspace, only_up = false, proto = nil, addresses = nil, ports = nil, names = nil)
+  def services(opts)
+    opts.delete(:workspace) # Mdm::Service apparently doesn't have an upstream Mdm::Workspace association
+
+  ::ActiveRecord::Base.connection_pool.with_connection {
+    search_term = opts.delete(:search_term)
+    only_up = opts.delete(:only_up)
+    opts[:state] = [Msf::ServiceState::Open] if only_up
+    opts["hosts.address"] = opts.delete(:addresses)
+
+    opts.compact!
+    column_search_conditions = Msf::Util::DBManager.create_all_column_search_conditions(Mdm::Service, search_term)
+    Mdm::Service.includes(:host).where(opts).where(column_search_conditions).order("hosts.address, port")
```

Comment:
I think symbols for ordering might be preferred to string. `order(Mdm::Host.arel_table[:address], :port)`

Reply:
It blows up when I try this.

```
msf5 > services
[-] Error while running command services: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "hosts"
LINE 1: ...T("services"."info" AS TEXT) ~ '(?mi)')  ORDER BY "hosts"."a...
                                                             ^
: SELECT "services".* FROM "services" WHERE (CAST("services"."id" AS TEXT) ~ '(?mi)' OR CAST("services"."host_id" AS TEXT) ~ '(?mi)' OR CAST("services"."created_at" AS TEXT) ~ '(?mi)' OR CAST("services"."port" AS TEXT) ~ '(?mi)' OR CAST("services"."proto" AS TEXT) ~ '(?mi)' OR CAST("services"."state" AS TEXT) ~ '(?mi)' OR CAST("services"."name" AS TEXT) ~ '(?mi)' OR CAST("services"."updated_at" AS TEXT) ~ '(?mi)' OR CAST("services"."info" AS TEXT) ~ '(?mi)')  ORDER BY "hosts"."address", "services"."port" ASC
```